ring/model_test.go

```diff
@@ -171,6 +172,102 @@ func TestDesc_Ready(t *testing.T) {
 	}
 }
 
+func TestDesc_ReadyZone(t *testing.T) {
+	now := time.Now()
+
+	r := &Desc{
+		Ingesters: map[string]InstanceDesc{
+			"ing1": {
+				Tokens:    []uint32{100, 200, 300},
+				State:     ACTIVE,
+				Timestamp: now.Unix(),
+				Zone:      "a",
+			},
+		},
+	}
+
+	expectedZone := "a"
+	t.Run("default cases", func(t *testing.T) {
+		if err := r.IsReadyZoneAware(now, 10*time.Second, expectedZone); err != nil {
+			t.Fatal("expected ready, got", err)
+		}
+
+		if err := r.IsReadyZoneAware(now, 0, expectedZone); err != nil {
+			t.Fatal("expected ready, got", err)
+		}
+
+		require.ErrorIs(t, r.IsReadyZoneAware(now.Add(5*time.Minute), 10*time.Second, expectedZone), ErrHeartbeat)
+
+		require.NoError(t, r.IsReadyZoneAware(now.Add(5*time.Minute), 0, expectedZone), "expected ready (no heartbeat but timeout disabled)")
+
+		r = &Desc{
+			Ingesters: map[string]InstanceDesc{
+				"ing1": {
+					State:     ACTIVE,
+					Timestamp: now.Unix(),
+					Zone:      "a",
+				},
+			},
+		}
+
+		require.Error(t, r.IsReadyZoneAware(now, 10*time.Second, expectedZone), "expected !ready (no tokens), but got no error")
+
+		r.Ingesters["some ingester"] = InstanceDesc{
+			Tokens:    []uint32{12345},
+			Timestamp: now.Unix(),
+			Zone:      "b",
+		}
+		require.NoError(t, r.IsReadyZoneAware(now, 10*time.Second, expectedZone), "expected ready")
+	})
+
+	// An inactive ingester in the same zone is still valid
+	t.Run("inactive ingester in same zone", func(t *testing.T) {
+		r.Ingesters["some inactive ingester"] = InstanceDesc{
+			Tokens:    []uint32{12345},
+			Timestamp: now.Unix(),
+			Zone:      "a",
+			State:     LEAVING,
+		}
+
+		require.NoError(t, r.IsReadyZoneAware(now, 10*time.Second, expectedZone), "expected ready")
+	})
+
+	// An inactive ingester in a different zone is invalid
+	t.Run("inactive ingester in different zone", func(t *testing.T) {
+		r.Ingesters["some inactive ingester different zone"] = InstanceDesc{
+			Tokens:    []uint32{12345},
+			Timestamp: now.Unix(),
+			Zone:      "b",
+			State:     LEAVING,
+		}
+
+		require.Error(t, r.IsReadyZoneAware(now, 10*time.Second, expectedZone), "expected !ready (inactive ingester in different zone), but got no error")
+	})
+
+	// An inactive ingester in a same zone due to heartbeat is invalid
+	t.Run("inactive ingester in same zone, heartbeat fails", func(t *testing.T) {
```

Comment:
This sounds like a different concern (and could be a different PR/etc), but shouldn't an ingester in the same zone failing heartbeat _not_ prevent a new ingester from joining?